### PR TITLE
spirv-fuzz: Fuzzer pass to add equation instructions

### PR DESCRIPTION
**Pull request by afd**
This introduces a new fuzzer pass to add instructions to the module
that define equations, and support in the fact manager for recording
equation facts and deducing synonym facts from equation facts.

Initially the only equations that are supported involve OpIAdd,
OpISub, OpSNegate and OpLogicalNot, but there is scope for adding
support for equations over various other operators.

### CODE REVIEW

File: source/fuzz/fuzzer_pass_apply_id_synonyms.cpp [hunk 1]
```diff
@@ -86,7 +86,8 @@ void FuzzerPassApplyIdSynonyms::Apply() {
         synonyms_to_try.erase(synonyms_to_try.begin() + synonym_index);
 
         if (synonym_to_try->index_size() > 0 &&
-            use_inst->opcode() == SpvOpPhi) {
+            !fuzzerutil::CanInsertOpcodeBeforeInstruction(SpvOpCompositeExtract,
+                                                          use_inst)) {
```
Comment (paulthomson) [inline source/fuzz/fuzzer_pass_apply_id_synonyms.cpp]: Update comment below

Comment (paulthomson) [reply source/fuzz/fuzzer_pass_apply_id_synonyms.cpp]: And add comment about index_size() > 0

File: source/fuzz/fuzzer_pass_apply_id_synonyms.cpp [hunk 2]
```diff
@@ -111,32 +112,16 @@ void FuzzerPassApplyIdSynonyms::Apply() {
...
-              synonym_to_try->object(),
-              fuzzerutil::RepeatedFieldToVector(synonym_to_try->index()));
-          assert(composite_extract_transformation.IsApplicable(
-                     GetIRContext(), *GetFactManager()) &&
-                 "Transformation should be applicable by construction.");
-          composite_extract_transformation.Apply(GetIRContext(),
-                                                 GetFactManager());
-          *GetTransformations()->add_transformation() =
-              composite_extract_transformation.ToMessage();
+          ApplyTransformation(TransformationCompositeExtract(
```
Comment (paulthomson) [inline source/fuzz/fuzzer_pass_apply_id_synonyms.cpp]: Could add comment above for index_size again

File: source/fuzz/transformation_equation_instruction.cpp [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+            (operand_type->AsVector() &&
+             operand_type->AsVector()->element_type()->AsInteger()))) {
+        return 0;
+      }
+      return operand_inst->type_id();
+    }
+
+    default:
+      // We do not know what to do with an equation that uses this opcode.
+      return 0;
```
Comment (paulthomson) [inline source/fuzz/transformation_equation_instruction.cpp]: assert(false)?

File: source/fuzz/fact_manager.cpp [hunk 4]
```diff
@@ -396,41 +481,252 @@ class FactManager::DataSynonymFacts {
   // whether a new fact has been added since the last time such a computation
   // was performed.
   //
-  // It is mutable so faciliate having const methods, that provide answers to
+  // It is mutable so facilitate having const methods, that provide answers to
```
[the remark this answers is not in the document]
Comment (afd) [reply source/fuzz/fact_manager.cpp]: Fixed.

File: source/fuzz/fact_manager.cpp [hunk 5]
```diff
@@ -396,41 +481,252 @@ class FactManager::DataSynonymFacts {
...
+    assert(synonymous_.Exists(*rhs_dd));
+    assert(synonymous_.Find(rhs_dd) == rhs_dd);
+  }
+
+  if (id_equations_.count(&lhs_dd) == 0) {
+    // We have not seen an equation with this LHS before, so associate the LHS
+    // with an initially empty set.
+    id_equations_.insert(
+        {&lhs_dd,
+         std::unordered_set<Operation, OperationHash, OperationEquals>()});
```
[the remark this answers is not in the document]
Comment (afd) [reply source/fuzz/fact_manager.cpp]: Yes, done!

File: source/fuzz/fuzzer_pass_add_equation_instructions.cpp [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+    opt::IRContext* ir_context, FactManager* fact_manager,
+    FuzzerContext* fuzzer_context,
+    protobufs::TransformationSequence* transformations)
+    : FuzzerPass(ir_context, fact_manager, fuzzer_context, transformations) {}
+
+FuzzerPassAddEquationInstructions::~FuzzerPassAddEquationInstructions() =
+    default;
+
+void FuzzerPassAddEquationInstructions::Apply() {
+  MaybeAddTransformationBeforeEachInstruction(
```
[the remark this answers is not in the document]
Comment (afd) [reply source/fuzz/fuzzer_pass_add_equation_instructions.cpp]: :) indeed.

File: source/fuzz/fuzzer_pass_apply_id_synonyms.cpp (hunk 1 above)
Comment (afd) [reply source/fuzz/fuzzer_pass_apply_id_synonyms.cpp]: Done.

File: source/fuzz/fuzzer_pass_apply_id_synonyms.cpp (hunk 2 above)
Comment (afd) [reply source/fuzz/fuzzer_pass_apply_id_synonyms.cpp]: Done.

File: source/fuzz/transformation_equation_instruction.cpp (hunk 3 above)
Comment (afd) [reply source/fuzz/transformation_equation_instruction.cpp]: Agreed.

**Pull request merged**